drivers/ethernet/dsa_ksz8794.c

```diff
@@ -0,0 +1,413 @@
+
+#define LOG_MODULE_NAME dsa
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <device.h>
+#include <kernel.h>
+#include <sys/util.h>
+#include <drivers/spi.h>
+#include <net/ethernet.h>
+#include <linker/sections.h>
+
+#include "dsa_ksz8794.h"
+
+#define DSA_KSZ8794_SLAVE_PORT_NUM \
+	DT_PROP(DT_CHILD(DT_NODELABEL(spi1), ksz8794_0), dsa_slave_ports)
+
+#define DSA_KSZ8794_MASTER_PORT_NODE \
+	DT_PHANDLE(DT_CHILD(DT_NODELABEL(spi1), ksz8794_0), dsa_master_port)
+
+static struct dsa_ksz8794_spi phy_spi;
+static struct spi_cs_control cs_con;
+
+static void dsa_ksz8794_write_reg(struct dsa_ksz8794_spi *sdev,
+				  u16_t reg_addr, u8_t value)
+{
+	u8_t buf[3];
+
+	const struct spi_buf tx_buf = {
+		.buf = buf,
+		.len = 3
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+
+	buf[0] = KSZ8794_SPI_CMD_WR | ((reg_addr >> 7) & 0x1F);
+	buf[1] = (reg_addr << 1) & 0xFE;
+	buf[2] = value;
+
+	spi_write(sdev->spi, &sdev->spi_cfg, &tx);
+}
+
+static void dsa_ksz8794_read_reg(struct dsa_ksz8794_spi *sdev,
+				 u16_t reg_addr, u8_t *value)
+{
+	u8_t buf[3];
+
+	const struct spi_buf tx_buf = {
+		.buf = buf,
+		.len = 3
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+	struct spi_buf rx_buf = {
+		.buf = buf,
+		.len = 3
+	};
+
+	const struct spi_buf_set rx = {
+		.buffers = &rx_buf,
+		.count = 1
+	};
+
+	buf[0] = KSZ8794_SPI_CMD_RD | ((reg_addr >> 7) & 0x1F);
+	buf[1] = (reg_addr << 1) & 0xFE;
+	buf[2] = 0x0;
+
+	if (!spi_transceive(sdev->spi, &sdev->spi_cfg, &tx, &rx)) {
+		*value = buf[2];
+	} else {
+		LOG_DBG("Failure while reading register 0x%04x", reg_addr);
+		*value = 0U;
+	}
+
+}
+
+static bool dsa_ksz8794_port_link_status(struct dsa_ksz8794_spi *sdev,
+					 u8_t port)
+{
+	u8_t tmp;
+
+	if(port < KSZ8794_PORT1 || port > KSZ8794_PORT3)
+		return false;
+
+	dsa_ksz8794_read_reg(sdev, KSZ8794_STAT2_PORTn(port), &tmp);
+
+	return tmp & KSZ8794_STAT2_LINK_GOOD;
+}
+
+static bool dsa_ksz8794_link_status(struct dsa_ksz8794_spi *sdev)
+{
+	bool ret = false;
+	u8_t i;
+
+	for (i = KSZ8794_PORT1; i <= KSZ8794_PORT3; i++) {
+		if (dsa_ksz8794_port_link_status(sdev, i)) {
+			LOG_INF("Port: %d link UP!", i);
+			ret |= true;
+		}
+	}
+
+	return ret;
+}
+
+static void dsa_ksz8794_soft_reset(struct dsa_ksz8794_spi *sdev)
+{
+	/* reset switch */
+	dsa_ksz8794_write_reg(sdev, KSZ8794_PD_MGMT_CTRL1,
+			      KSZ8794_PWR_MGNT_MODE_SOFT_DOWN);
+	k_busy_wait(1000);
+	dsa_ksz8794_write_reg(sdev, KSZ8794_PD_MGMT_CTRL1, 0);
+}
+
+static int dsa_ksz8794_spi_setup(struct dsa_ksz8794_spi *sdev, u32_t freq,
+				 u8_t cs_num, char *cs_gpio_name,
+				 u8_t cs_gpio_num)
+{
+	u16_t timeout = 100;
+	u8_t val[2], tmp;
+
+	/* SPI config */
+	sdev->spi_cfg.operation = SPI_WORD_SET(8) | SPI_MODE_CPOL
+		| SPI_MODE_CPHA;
+
+	sdev->spi_cfg.frequency = freq;
+	sdev->spi_cfg.slave = cs_num;
+
+	cs_con.gpio_dev = device_get_binding(cs_gpio_name);
```

Comment:
I don't know anything about this hw, but are you sure the CS will be always handled via a gpio line? What if the cs is handled by the SPI controller internally?

Reply:
The KSZ8794 in the switch configuration uses the SPI for communication.
The rationale for explicitly using "cs_gpio" is the lack of support for CS controlled by NXP's SPI IP block directly.
However, I do agree that the code in the dsa_ksz8794 shall be written in a way, that other SoCs could reuse it (and maybe those would support in-hardware control of CS).